File: src/simulation/space/resource/Resource.java
package simulation.space.resource;

import extra.SpaceProbabilityFuncs;
import kotlin.Pair;
import simulation.Event;
import simulation.culture.aspect.Aspect;
import simulation.culture.aspect.AspectResult;
import simulation.culture.thinking.meaning.Meme;
import simulation.space.resource.tag.AspectImprovementTag;
import simulation.space.tile.Tile;
import simulation.space.resource.dependency.ResourceDependency;
import simulation.space.resource.tag.ResourceTag;
import simulation.space.tile.TileTag;

import java.util.*;

import static shmp.random.RandomProbabilitiesKt.*;
import static simulation.Controller.session;

public class Resource {
    protected int amount;
    ResourceCore resourceCore;

    /**
     * Precomputed hash.
     */
    private int _hash;
    /**
     * How many turns has this Resource been existing.
     */
    private int deathTurn = 0;
    /**
     * How many additional years added to this Resource due to bad environment.
     * Large numbers results in sooner death.
     */
    private int deathOverhead = 0;
    /**
     * What part of this Resource will be destroyed on the next death.
     */
    private double deathPart = 1;
    private List<Event> events = new ArrayList<>();
    private OwnershipMarker ownershipMarker = OwnershipMarkerKt.getFreeMarker();

    Resource(ResourceCore resourceCore, int amount) {
        this.amount = amount;
        this.resourceCore = resourceCore;
        computeHash();
        events.add(new Event(Event.Type.Creation, "Resource was created", "name", getFullName()));
    }

    public Resource(ResourceCore resourceCore) {
        this(resourceCore, resourceCore.getGenome().getDefaultAmount());
    }

    public Map<Aspect, List<Pair<Resource, Integer>>> getAspectConversion() {
        return resourceCore.getAspectConversion();
    }

    void computeHash() {
        _hash = Objects.hash(getFullName());
    }

    public String getSimpleName() {
        return getGenome().getName();
    }

    public String getBaseName() {
        return getGenome().getBaseName();
    }

    public String getFullName() {
        return getGenome().getBaseName() + resourceCore.getMeaningPostfix();
    }

    public int getAmount() {
        return amount;
    }

    public List<ResourceTag> getTags() {
        return getGenome().getTags();
    }

    public double getAspectImprovement(Aspect aspect) {
        return getTags().stream()
                .filter(t -> t instanceof AspectImprovementTag && ((AspectImprovementTag) t).getLabeler().isSuitable(aspect))
                .map(t -> ((AspectImprovementTag) t).getImprovement())
                .reduce(0.0, Double::sum) * amount  ;
    }

    public double getTagPresence(ResourceTag tag) {
        return amount * getTagLevel(tag) * getGenome().getSize();
    }

    public int getTagLevel(ResourceTag tag) {
        int index = getTags().indexOf(tag);
        return index == -1 ? 0 : getTags().get(index).level;
    }

    /**
     * Returns part of this resource and subtracts its amount from this resource amount;
     * @param part what amount of this Resource is requested.
     * @return Copy of this Resource with amount equal or less than requested.
     * Exact amount depends on current amount of this Resource.
     */
    public Resource getPart(int part) {
        int result;
        double prob = session.random.nextDouble() * 0.5;
        if (part <= amount * prob) {
            result = Math.min(amount, part);
        } else {
            result = amount * prob + 1 < amount ? (int) (amount * prob) + 1 : amount;
        }
        amount -= result;
        return copy(result);
    }

    public Resource getCleanPart(int part) {
        int result = Math.min(amount, part);
        amount -= result;
        return copy(result);
    }

    public Genome getGenome() {
        return resourceCore.getGenome();
    }

    public OwnershipMarker getOwnershipMarker() {
        return ownershipMarker;
    }

    public void setOwnershipMarker(OwnershipMarker ownershipMarker) {
        this.ownershipMarker = ownershipMarker;
    }

    public boolean hasMeaning() {
        return resourceCore.getHasMeaning();
    }

    public void setHasMeaning(boolean b) {
        resourceCore.setHasMeaning(b);
    }

    public Resource merge(Resource resource) {
        if (!resource.getBaseName().equals(getBaseName())) {
            throw new RuntimeException(String.format(
                    "Different resource tried to merge - %s and %s",
                    getFullName(),
                    resource.getFullName()
            ));
        }
        if (this == resource) {
            return this;
        }
        addAmount(resource.amount);
        resource.destroy();
        return this;
    }

    public Resource copy() {
        Resource resource = resourceCore.copy();
        resource.setOwnershipMarker(ownershipMarker);
        return resource;
    }

    public Resource copy(int amount) {
        Resource resource = resourceCore.copy(amount);
        resource.setOwnershipMarker(ownershipMarker);
        return resource;
    }

    public Resource fullCopy() {
        Resource resource = resourceCore.fullCopy();
        resource.setOwnershipMarker(ownershipMarker);
        return resource;
    }

    public Resource insertMeaning(Meme meaning, AspectResult result) {
        Resource resource = new Resource(resourceCore.insertMeaning(meaning, result), amount);
        destroy();
        return resource;
    }

    public ResourceUpdateResult update(Tile tile) {
        List<Resource> result = new ArrayList<>();
        if (amount <= 0) {
            return new ResourceUpdateResult(false, result);
        }
        for (ResourceDependency dependency: resourceCore.getGenome().getDependencies()) {
            double part = dependency.satisfactionPercent(tile, this);
            deathOverhead += (1 - part) * resourceCore.getGenome().getDeathTime();
        }
        if (deathTurn + deathOverhead >= resourceCore.getGenome().getDeathTime()) {
            int deadAmount = (int) (deathPart*amount);
            amount -= deadAmount;
            deathTurn = 0;
            deathOverhead = 0;
            deathPart = 1;
            result = applyAspect(ResourcesKt.getDEATH_ASPECT(), deadAmount);
        }
        if (amount <= 0) {
            new ResourceUpdateResult(false, result);
        }
        deathTurn++;
        if (testProbability(resourceCore.getGenome().getSpreadProbability(), session.random)) {
            expand(tile);
        }
        if (getSimpleName().equals("Vapour")) {
            if (amount > 100000) {//TODO debug off
//                if (tile.getType() != Tile.Type.Water &&
//                        tile.getResource("Water").getAmount() < 5) {
//                    tile.addDelayedResource(session.world.getPoolResource("Water").copy(amount / 100));
//                    amount /= 100;
//                }
            }
            if (tile.getTemperature() < 0) {
                tile.addDelayedResource(session.world.getResourcePool().get("Snow").copy(amount / 2));
                amount -= amount / 2;
            }
        }
        distribute(tile);
        return new ResourceUpdateResult(true, result);
    }

    private void distribute(Tile tile) {
        if (getGenome().getCanMove() && amount > getGenome().getNaturalDensity()) {
            List<Tile> tiles = tile.getNeighbours(t -> getGenome().isAcceptable(t));
            tiles.sort(Comparator.comparingInt(t -> t.getResourcePack().getAmount(this)));
            for (Tile neighbour: tiles) {
                if (amount <= getGenome().getNaturalDensity() / 2) {
                    break;
                }
                int part = Math.min(amount - getGenome().getNaturalDensity() / 2,
                        getGenome().getNaturalDensity() - neighbour.getResourcePack().getAmount(this));
                part = part <= 0 ? (amount - getGenome().getNaturalDensity() / 2) / tiles.size() : part;
                neighbour.addDelayedResource(getCleanPart(part));
            }
        }
    }

    public void addAmount(int amount) {
        if (amount > 0) {
            deathPart = (this.amount * deathPart) / (this.amount + amount);
        }
        this.amount += amount;
    }

    public List<Resource> applyAspect(Aspect aspect, int part) {
        List<Resource> result = resourceCore.applyAspect(aspect);
        result.forEach(r -> r.amount *= part);
        return result;
    }

    public List<Resource> applyAspect(Aspect aspect) {
        return applyAspect(aspect, 1);
    }

    public boolean hasApplicationForAspect(Aspect aspect) {
        return resourceCore.hasApplicationForAspect(aspect);
    }

    public void destroy() {
        this.amount = 0;
    }

    public List<Resource> applyAndConsumeAspect(Aspect aspect, int part) {
        Resource _r = getPart(part);
        int p = Math.min(part, _r.amount);
        List<Resource> result = _r.resourceCore.applyAspect(aspect);
        result.forEach(resource -> resource.amount *= p);
        _r.amount -= p;
        return result;
    }

    private boolean expand(Tile tile) {
        List<Tile> l = new ArrayList<>();
        l.add(tile);
        Tile newTile = SpaceProbabilityFuncs.randomTileOnBrink(l, t -> getGenome().isAcceptable(t)
                && getGenome().getDependencies().stream().allMatch(dependency -> dependency.hasNeeded(t)));
        if (newTile == null) {
            if (getGenome().getDependencies().stream().allMatch(dependency -> dependency.hasNeeded(tile))) {
                newTile = tile;
            } else {
                newTile = SpaceProbabilityFuncs.randomTileOnBrink(l, t -> getGenome().isAcceptable(t));
                if (newTile == null) {
                    newTile = tile;
                }
            }
        }
        Resource resource = copy();
        resource.amount = Math.min(resourceCore.getGenome().getDefaultAmount(), amount);
        newTile.addDelayedResource(resource);
        return true;
    }

    @Override
    public boolean equals(Object o) {//TODO maybe re
        return fullEquals(o);
    }

    public boolean ownershiplessEquals(Object o) {
        if (this == o) return true;
        if (o == null) return false;
        Resource resource = (Resource) o;
        if (_hash != resource._hash) {
            return false;
        }
        return getFullName().equals(resource.getFullName());
    }

    public boolean fullEquals(Object o) {
        if (this == o) return true;
        if (o == null) return false;
        Resource resource = (Resource) o;
        if (_hash != resource._hash) {
            return false;
        }
        if (getFullName().equals(resource.getFullName()) && !ownershipMarker.equals(resource.ownershipMarker)) {
            int j = 0;
        }
        return getFullName().equals(resource.getFullName()) && ownershipMarker.equals(resource.ownershipMarker);
    }

    @Override
    public int hashCode() {
        return Objects.hash(getFullName(), resourceCore.hashCode(), ownershipMarker);
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder("Resource " + getFullName() + ", natural density - " +
                getGenome().getNaturalDensity() + ", spread probability - " + getGenome().getSpreadProbability() +
                ", mass - " + getGenome().getMass() + ", amount - " + amount + ", tags: ");
        for (ResourceTag resourceTag : getTags()) {
            stringBuilder.append(resourceTag.name).append(" ");
        }
        return stringBuilder.toString();
    }
}

File: src/simulation/space/resource/Resource.kt
package simulation.space.resource

import shmp.random.randomTileOnBrink
import shmp.random.testProbability
import simulation.Controller
import simulation.event.Event
import simulation.SimulationException
import simulation.space.SpaceData
import simulation.space.resource.action.ResourceAction
import simulation.space.resource.action.ResourceProbabilityAction
import simulation.space.resource.tag.ResourceTag
import simulation.space.tile.Tile
import java.util.*
import kotlin.math.min
import kotlin.math.pow

open class Resource(
        var core: ResourceCore,
        open var amount: Int = core.genome.defaultAmount,
        val ownershipMarker: OwnershipMarker = freeMarker
) {
    // Precomputed hash.
    private var _hash = 0

    //How many turns has this Resource been existing.
    private var deathTurn = 0

    //How many additional years added to this Resource due to bad environment. Large numbers results in sooner death.
    private var deathOverhead = 0

    //What part of this Resource will be destroyed on the next death.
    private var deathPart = 1.0
    private val events: MutableList<Event> = ArrayList()

    val isEmpty: Boolean
        get() {
            if (amount < 0)
                throw SimulationException("Resource amount is below zero: $this")
            return amount == 0
        }

    val isNotEmpty: Boolean
        get() = !isEmpty

    fun computeHash() {
        _hash = Objects.hash(fullName, core.hashCode(), ownershipMarker)
    }

    val simpleName: String
        get() = genome.name

    val baseName: String
        get() = genome.baseName

    val fullName: String
        get() = genome.baseName + if (externalFeatures.isNotEmpty())
            externalFeatures.joinToString("_", "_") { it.name }
        else ""

    val tags: List<ResourceTag>
        get() = genome.tags

    val externalFeatures = core.externalFeatures

    init {
        computeHash()
        events.add(Event(Event.Type.Creation, "Resource was created", "name", fullName))
    }

    fun getTagPresence(tag: ResourceTag) =
            (amount * getTagLevel(tag)).toDouble() * genome.size.pow(SpaceData.data.resourceSizeEffect)

    fun getTagLevel(tag: ResourceTag) = tags.firstOrNull { it == tag }?.level ?: 0

    /**
     * Returns part of this resource and subtracts its amount from this resource amount;
     * @return Copy of this Resource with amount equal or less than requested.
     * Exact amount depends on current amount of this Resource.
     */
    open fun getPart(part: Int): Resource {
        val prob = SpaceData.data.random.nextDouble() * 0.5
        val result = when {
            part <= amount * prob -> min(amount, part)
            amount * prob + 1 < amount -> (amount * prob).toInt() + 1
            else -> amount
        }
        amount -= result

        return copy(result)
    }

    fun getCleanPart(part: Int): Resource {
        val result = min(amount, part)
        amount -= result
        return copy(result)
    }

    val genome: Genome
        get() = core.genome

    open fun merge(resource: Resource): Resource {
        if (resource.baseName != baseName) throw RuntimeException(String.format(
                "Different resource tried to merge - %s and %s",
                fullName,
                resource.fullName
        ))
        if (this === resource)
            return this

        addAmount(resource.amount)
        resource.destroy()
        return this
    }

    fun exactCopy(ownershipMarker: OwnershipMarker = this.ownershipMarker) = copy(amount, ownershipMarker)

    fun exactCopyAndDestroy(ownershipMarker: OwnershipMarker = this.ownershipMarker) =
            copyAndDestroy(amount, ownershipMarker)

    fun copy(amount: Int = genome.defaultAmount, ownershipMarker: OwnershipMarker = this.ownershipMarker) =
            Resource(core, amount, ownershipMarker)

    fun copyAndDestroy(
            amount: Int = genome.defaultAmount,
            ownershipMarker: OwnershipMarker = this.ownershipMarker
    ): Resource {
        val result = Resource(core, amount, ownershipMarker)
        destroy()
        return result
    }

    fun fullCopy() = core.fullCopy(ownershipMarker)

    fun copyWithExternalFeatures(features: List<ExternalResourceFeature>): Resource {
        val resource = Resource(core.copyWithNewExternalFeatures(features), amount)
        destroy()
        return resource
    }

    fun copyWithNewExternalFeatures(features: List<ExternalResourceFeature>) = copyWithExternalFeatures(externalFeatures + features)

    open fun update(tile: Tile): ResourceUpdateResult {
        val result = mutableListOf<Resource>()
        if (amount <= 0)
            return ResourceUpdateResult(false, result)

        result.addAll(genome.conversionCore.probabilityActions.flatMap { applyProbabilityAction(it) })

        for (dependency in core.genome.dependencies) {
            val part = dependency.satisfactionPercent(tile, this)
            deathOverhead += ((1 - part) * core.genome.lifespan).toInt()
        }

        if (deathTurn + deathOverhead >= core.genome.lifespan) {
            val deadAmount = (deathPart * amount).toInt()
            amount -= deadAmount
            deathTurn = 0
            deathOverhead = 0
            deathPart = 1.0
            result.addAll(applyAction(specialActions.getValue("_OnDeath_"), deadAmount))
        }

        if (amount <= 0)
            ResourceUpdateResult(false, result)
        deathTurn++
        if (testProbability(core.genome.spreadProbability, SpaceData.data.random))
            expand(tile)

        if (simpleName == "Vapour") {
            if (tile.temperature < 0) {
                tile.addDelayedResource(Controller.session.world.resourcePool.get("Snow").copy(amount / 2))
                amount -= amount / 2
            }
        }

        distribute(tile)

        return ResourceUpdateResult(true, result)
    }

    private fun applyProbabilityAction(action: ResourceProbabilityAction): List<Resource> {
        val expectedValue = amount * action.probability
        val part =
                if (expectedValue < 1.0)
                    if (testProbability(expectedValue, Controller.session.random))
                        1
                    else 0
                else expectedValue.toInt()

        return if (action.isWasting)
            applyActionAndConsume(action, part, true)
        else
            applyAction(action, part)
    }

    fun isAcceptable(tile: Tile) =
            genome.dependencies.filter { it.isNecessary }.all { it.satisfactionPercent(tile, this) == 1.0 }

    fun isOptimal(tile: Tile) = isAcceptable(tile)
            && genome.dependencies.filter { !it.isPositive }.all { it.satisfactionPercent(tile, this) >= 0.9 }

    private fun distribute(tile: Tile) {
        if (amount <= genome.naturalDensity)
            return

        when (genome.overflowType) {
            OverflowType.Migrate -> {
                val tiles = tile.getNeighbours { isAcceptable(it) }
                        .sortedBy { it.resourcePack.getAmount(this) }

                for (neighbour in tiles) {
                    if (amount <= genome.naturalDensity / 2)
                        break

                    var part = min(
                            amount - genome.naturalDensity / 2,
                            genome.naturalDensity - neighbour.resourcePack.getAmount(this)
                    )
                    part = if (part <= 0)
                        (amount - genome.naturalDensity / 2) / tiles.size
                    else part

                    neighbour.addDelayedResource(getCleanPart(part))
                }
            }
            OverflowType.Cut -> amount = genome.naturalDensity
            OverflowType.Ignore -> {}
        }
    }

    open fun addAmount(amount: Int) {
        if (amount > 0)
            deathPart = this.amount * deathPart / (this.amount + amount)
        this.amount += amount
    }

    fun applyAction(action: ResourceAction, part: Int = 1): List<Resource> {
        val result = core.applyAction(action)
        result.forEach { it.amount *= part }
        return result
    }

    fun hasApplicationForAction(action: ResourceAction) = genome.conversionCore.hasApplication(action)

    fun destroy() {
        amount = 0
    }

    open fun applyActionAndConsume(action: ResourceAction, part: Int, isClean: Boolean): List<Resource> {
        val resourcePart = if (isClean)
            getCleanPart(part)
        else
            getPart(part)

        val p = min(part, resourcePart.amount)
        val result = resourcePart.core.applyAction(action)
        result.forEach { it.amount *= p }
        resourcePart.amount -= p

        return result
    }

    private fun expand(tile: Tile): Boolean {
        val tileList = mutableListOf(tile)

        var newTile = randomTileOnBrink(tileList, SpaceData.data.random) { t ->
            (isAcceptable(t) && genome.dependencies.all { d -> d.hasNeeded(t) })
        }
        if (newTile == null) {
            if (genome.dependencies.all { d -> d.hasNeeded(tile) })
                newTile = tile
            else {
                newTile = randomTileOnBrink(tileList, SpaceData.data.random) { isAcceptable(it) }
                if (newTile == null)
                    newTile = tile
            }
        }
        val resource = copy()
        resource.amount = min(core.genome.defaultAmount, amount)
        newTile.addDelayedResource(resource)
        return true
    }

    override fun equals(other: Any?) = fullEquals(other)

    fun ownershiplessEquals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null) return false
        val resource = o as Resource

        return if (_hash == resource._hash)
            fullName == resource.fullName
        else false
    }

    fun fullEquals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null) return false
        val resource = o as Resource
        return if (_hash != resource._hash)
            false
        else
            fullName == resource.fullName && ownershipMarker == resource.ownershipMarker
    }

    override fun hashCode() = _hash

    override fun toString() = "Resource $fullName, natural density - ${genome.naturalDensity}" +
            ", spread probability - ${genome.spreadProbability}, mass - ${genome.mass}, " +
            "lifespan - ${genome.lifespan}, amount - $amount, ownership - $ownershipMarker, tags: " +
            tags.joinToString(" ") { it.name }
}
